docs/proposals/repo-server-multi-tenancy.md

```diff
@@ -0,0 +1,200 @@
+
+If the `.spec.repoServer` field is specified, both `argocd-server` and `argocd-application-controller` would look up the appropriate configuration for the repository server instance to use, and create a gRPC clientset specifically for this instance when submitting requests to the repository server. When the referenced configuration is not found, there is no fallback to the default one. Instead, an error would be produced and manifest generation will fail. This is important for security reasons, so that a typo or misconfiguration would not lead to a potential flaw in isolation.
+
+The relationship of an `AppProject` to a repository server instance is `1:n`. An `AppProject` instance can use exactly one repository server instance, while each repository server instance can be referenced by multiple `AppProjects`. 
+
+### Changes to caching
+
+Redis is used as a central instance to cache manifests generated by the repository server. The manifest cache currently uses keys that do not consider multiple independent repository servers storing manifests in the same cache. Also, by default there is no authentication between the repository server and the cache, and access control is solely implemented using network policies.
+
+There are a couple of challenges to be solved with the cache:
+
+1. Repository server instances must be able to make distinct cache entries, possibly for the same repositories and same revisions.
+1. Any repository server instance must not be allowed to gain access to cache entries from another repository server instance
+
+In initial discussions, a possible solution for the first challenge would be to prefix cache keys with the instance name of the repository server (e.g. `instance-1`). This is easy to implement, and only requires minimal changes and efforts to the code.
+
+For the second challenge however, there have been multiple ideas of various complexity.
+
+**1. Using authentication and ACLs in Redis**: This would involve setting up users and ACLs in Redis, and augmenting repository server instance configuration with credentials. This is however, complex to setup and maintain and could also prevent some users from using external (managed) Redis instance where this is not supported or would cost additional money.
+
+**2. Moving the manifest cache to the application controller:** Another idea is to move the handling of the cache completely to the application controller, and remove the relationship between repository server and the cache. This would also save some network traffic between the application controller and the repository server, which would be especially efficient when the application controller requests manifests from the repository server and the repository server would just return the cached manifests.
```

Comment:
Ah, good observation. That'd be a problem, because we would basically duplicate the cache on the API server then, right, and then we wouldn't even know if the cache on the application server is similar to the cache on the API server. Could create some serious split-brain-like scenario.